xgcm/test/test_metrics.py

```diff
@@ -283,23 +111,22 @@ def test_assign_metric(key, metric_vars):
         (["X", "Y", "Z"], "tracer", None, ["volume_t"], None),
         (["X"], "u", None, ["dx_e"], None),
         (["X", "Y"], "u", None, ["area_e"], None),
-        (("X", "Y", "Z"), "u", None, ["volume_t"], KeyError),  # This should error out
+        (("X", "Y", "Z"), "u", ["dz_t"], ["volume_t"], KeyError),  # This should error out
         # reconstructed cases
         (["X", "Y"], "tracer", ["area_t"], ["dx_t", "dy_t"], None),
         (["X", "Y", "Z"], "tracer", ["volume_t"], ["area_t", "dz_t"], None),
+        (["X", "Y", "Z"], "u", None, ["area_e", "dz_t"], None),
```

Comment:
I just added this as a reconstruct test case, without dropping `dz_t` because why not - sometimes those velocity points are tricky...

Reply:
Same thing here. This should not work! `dz_t` is located at `xt` and `yt`, the vertical data to recostruct the volume for `u` should be located at the same point as `area_e`.

Reply:
So this happens because of how I changed the `dz_t` and `dz_w` elements. Because they are now 1D vectors (just like in e.g. MITgcm) they are not tied to particular horizontal points, and therefore xgcm "reconstructs" volume at the `u` point with the product `area_e` and `dz_t`. Note that `area_e` * `dz_t` is not "equal to" `volume_t`, so if instead of having `area_e`,`dz_t` in the expected returned metrics, the test will error out. However, `area_e` * `dz_t` should be the correct volumetric element at the `u` point because the vertical grid spacing does not change from the tracer point on the horizontal grid to any of the velocity points. 

If you want me to make the `dz` elements so that they are instead 3D fields, defined at specific points on the horizontal grid, then I can do this. I just didn't do this because (1) the notation got pretty hairy and I had to make 6 different dz elements which seemed a bit unnecessary and (2) at least the MITgcm uses 1D vertical coordinates so it seemed like a more realistic case... Basically I made this change because I was able to get the same `KeyError` as you had written in `test_get_metric`, and the same behavior in all the other previous tests, and this seemed like an easier solution than making a bunch of 3D fields for `dz` elements. I'm open to being convinced otherwise though :)

Reply:
I see. Thanks for the explanation. I must have missed the changes in the `dz_t` (github does not offer an option to compare your PR to another (#130) as far as I know, which makes it difficult to see the exact changes).

I think your test case is very relevant, but I would prefer them to be implemented as another `grid_type`. E.g. `ds = dataset_grid_metric('C_fixed_dz`)` (maybe there is a better naming hehe). 
The reason why I want to test the most general case extensively (3D or actually 4D vertical spacing) is that many models use some sort of variable depth spacing for the cells (MOM5 uses z* coordinates and MOM6 has a completely arbitrary vertical coordinate system) and I believe xgcm can be tremendously useful for these grid layouts. 

We need to take care that our methods are not specific to a single model, but it is a great idea to test special cases as you suggested. 

My suggestion is: 
1) Create a separate dataset with a fixed vertical spacing. This should be achievable within ` dataset_grid_metric` relatively easy.
2) Add a corresponding test function `test_derivative_c_grid_fixed_z`/ `test_get_metric_c_grid_fixed_z` or similarly named.

That way all my test stay intact and we add the important case of 1d fixed vertical cell depth to the tests.

Reply:
Ok, I think that makes sense to have 3D/4D dz grid elements then. Thanks for all the explanation, I guess I should take another model for a spin (heh) to get a flavor for other grids. (It's always been interesting to me that MITgcm has the z* coordinate system but still the vertical grid is printed as a 1D, static vector). 

I'll be able to do this later today or this weekend, it should not be difficult as you said. Coming soon!

Reply:
Thats why we work as a team on this! Different perspectives give us a more general approach.

>It's always been interesting to me that MITgcm has the z* coordinate system but still the vertical grid is printed as a 1D, static vector

MOM5 does the same thing but the vertical cell depth is in fact changing in time (particularly near the surface), there is probably an option to output the 'real' `dz` as a diagnostic. It is also possible to reconstruct a z* coordinate from the SSH and ocean depth. I have implemented the reconstruction [here](https://github.com/jbusecke/mom_xtools/blob/928f780d59570d96c629e12d87d08b366ffbb024/mom_xtools/mom_budgets.py#L185) *Disclaimer*: This might be MOM specific. 
There are execellent (although not a short read) explanation on that topic in the [MOM5 docs](https://mom-ocean.github.io/assets/pdfs/MOM5_manual.pdf)

Reply:
Nice! Yeah this looks MOM specific, but with `xgcm.diff` it would not be difficult to implement for MITgcm diagnostics... 

As for the PR, I just added dz elements for velocity points to the `datasets_grid_metric`. Let me know what you think!